In [95]:
import pandas as pd
import numpy as np

df = pd.read_csv("https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding="latin-1")
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [ ]:
df.Item.unique()

In [ ]:
df.info()

# What is the total Protein supply quantity in Madagascar in 2015?

In [ ]:

madagascar_2015 = df[(df['Area'] == 'Madagascar') & (df['Element'] == 'Protein supply quantity (g/capita/day)')]
display(madagascar_2015)

In [25]:
total_protein_supply = madagascar_2015["Y2015"].sum()
display(total_protein_supply)

173.04999999999998

# Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [ ]:

df_selected = df[['Area', 'Y2017']]
display(df_selected)


In [ ]:
df_group = df_selected.groupby('Area')['Y2017'].sum()
display(df_group)

In [38]:
seventh_lowest = df_group.sort_values(ascending=True).index[6]
display(seventh_lowest)

'Guinea-Bissau'

# What is the total number of unique countries in the dataset

In [41]:

unique_countries = df.Area.unique()

len(unique_countries)

49

# What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [44]:

df_missingno = df['Y2014'].isnull().sum()
display(df_missingno)

1589

In [53]:
total_count_2014 = len(df["Y2014"])
display(total_count_2014)

60943

In [55]:
percent_of_missingval = (df_missingno/total_count_2014)*100
round(percent_of_missingval, 3)


2.607

# Which year had the least correlation with ‘Element Code’?

In [63]:
df.corr()

<ipython-input-63-2f6f6606aa2c>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

Answer :**Y2016** 

# Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [73]:
grp_element = df.groupby('Element')['Y2017'].sum()
display(grp_element)

Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64

Processing:                              292836.00

# What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [85]:
mean_Y2017 = df["Y2017"].mean()

round(mean_Y2017, 2)

140.92

In [86]:
std_Y2017 = df["Y2017"].std()

round(std_Y2017, 2)

1671.86

# Answer the following questions based on the African food production dataset provided by the FAO website already provided

# What is the total sum of Wine produced in 2015 and 2018 respectively?

# Hint:

# Perform a groupby sum aggregation on ‘Item’

In [94]:
wine = df[(df['Item'] == 'Wine')]

df_wine = wine.groupby('Item')['Y2015', 'Y2018'].sum()
display(df_wine)

<ipython-input-94-df9a668fcd5e>:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_wine = wine.groupby('Item')['Y2015', 'Y2018'].sum()


Y2015    Y2018
Item                  
Wine  4251.81  4039.32

# Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [ ]:
df_selected1 = df[['Area', 'Y2017']]
display(df_selected1)

In [ ]:
highest_area = df_selected1.groupby('Area')['Y2017'].sum()
display(highest_area)

In [107]:
sort_highest_area = highest_area.sort_values(ascending = False)
display(sort_highest_area)

Area
Nigeria                        1483268.23
Egypt                           866379.92
South Africa                    517590.54
Ethiopia                        448683.76
Morocco                         388495.36
Ghana                           337599.06
Algeria                         325644.27
United Republic of Tanzania     322616.85
Kenya                           264660.66
Sudan                           239931.92
Cameroon                        232030.43
Angola                          229159.57
Côte d'Ivoire                   224599.01
Uganda                          213950.38
Malawi                          181098.71
Mozambique                      161407.98
Mauritania                      156665.46
Mali                            149928.33
Madagascar                      131197.73
Niger                           126707.58
Benin                           124771.22
Tunisia                         124167.20
Zambia                          103223.77
Burkina Faso                 

Nigeria:                    1483268.23

# Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [109]:
stock = df[(df['Element'] == 'Stock Variation')]

df_stock = stock.groupby('Element')['Y2014','Y2015','Y2016', 'Y2017','Y2018'].sum()
display(df_stock)

<ipython-input-109-97f34d0bc590>:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_stock = stock.groupby('Element')['Y2014','Y2015','Y2016', 'Y2017','Y2018'].sum()


Y2014     Y2015     Y2016     Y2017     Y2018
Element                                                          
Stock Variation  58749.83  34910.99  33140.12  54316.91  20577.91

Y2014:	58749.83